### PR TITLE
Pcx 313 smartswitch

### PR DESCRIPTION
**Pull request by siebeprojects**
Remove smart switch logic from the NetworkCard and PaymentNetwork classes.
This makes it easier to re-use the PaymentNetwork class for GooglePay activation.

### CODE REVIEW

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.List;
+
+import android.text.TextUtils;
+
+/**
+ * Class for storing and selecting payment networks based on smart switch logic
+ */
+public final class SmartSwitch {
+
+    private final HashMap<String, String> smartMapping;
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: You should probably use `Map` here - that the instance is a `HashMap` is an implementation detail.

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        return smartSelected;
+    }
+
+    /**
+     * Get the first smart selected PaymentNetwork, may return null if none are selected.
+     *
+     * @return the first smart selected PaymentNetwork.
+     */
+    public PaymentNetwork getFirstSelected() {
+        return smartSelected.size() > 0 ? smartSelected.get(0) : null;
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: Minor: better use `isEmpty` (might make a difference with some list implementations)

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    /**
+     * Check if the PaymentNetwork is smart selected, it is smart selected when the provided number input matches
+     * the regex of this PaymentMethod in the groups settings file. A PaymentMethod is always smart selected when
+     * there is only one PaymentMethod in the NetworkCard.
+     *
+     * @param network to check if smart selection
+     * @return true when smart selected, false otherwise
+     */
+    public boolean isSelected(PaymentNetwork network) {
+        if (networks.size() == 1 && networks.get(0) == network) {
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: Almost surely an unnecessary optimization.

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    /**
+     * Validate the PaymentMethods for smart selection given the new text
+     *
+     * @param text to validate for smart selection
+     * @return true when the smart selection of payment methods has changed, false otherwise
+     */
+    public boolean validate(String text) {
+        smartBuffer.clear();
+        String regex;
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: I would prefer defining the variable where it is set - less clutter.

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+        if (text != null) {
+            for (PaymentNetwork network : networks) {
+                regex = smartMapping.get(network.getCode());
+
+                if (!TextUtils.isEmpty(regex) && text.matches(regex)) {
```
Comment (hjhill) [inline payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: Would pre-compiling the regex to a `Pattern` be better here? (for desktop Java it definitely would, but not sure about Android)

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java (hunk 3 above)
Comment (hjhill) [reply payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: Sorry, mixed up the two lists in my mind...

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java (hunk 1 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: Changed to Map

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java (hunk 2 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: Using isEmpty now

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java (hunk 4 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: Done

File: payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java (hunk 5 above)
Comment (siebeprojects) [reply payment/src/main/java/net/optile/payment/ui/model/SmartSwitch.java]: Using now Pattern instead of String regex

**Pull request merged**